service/dynamodb/dynamodbattribute/field.go

```diff
@@ -0,0 +1,267 @@
+package dynamodbattribute
+
+import (
+	"reflect"
+	"sort"
+	"strings"
+)
+
+type field struct {
+	tag
+
+	Name        string
+	NameFromTag bool
+
+	Index []int
+	Type  reflect.Type
+}
+
+func fieldByName(fields []field, name string) (field, bool) {
+	for _, f := range fields {
+		if f.Name == name {
+			return f, true
+		}
+	}
+	for _, f := range fields {
+		if strings.EqualFold(f.Name, name) {
+			return f, true
+		}
+	}
+
+	return field{}, false
+}
+
+func buildField(pIdx []int, i int, sf reflect.StructField, fieldTag tag) field {
+	f := field{
+		Name: sf.Name,
+		Type: sf.Type,
+		tag:  fieldTag,
+	}
+	if len(fieldTag.Name) != 0 {
+		f.NameFromTag = true
+		f.Name = fieldTag.Name
+	}
+
+	f.Index = make([]int, len(pIdx)+1)
+	copy(f.Index, pIdx)
+	f.Index[len(pIdx)] = i
+
+	return f
+}
+
+func unionStructFields(t reflect.Type, opts MarshalOptions) []field {
+	fields := enumFields(t, opts)
+
+	sort.Sort(fieldsByName(fields))
+
+	fields = visibleFields(fields)
+
+	return fields
+}
+
+// enumFields will recursively iterate through a structure and its nested
+// anonymous fields.
+//
+// Based on the enoding/json struct field enumeration of the Go Stdlib
+// https://golang.org/src/encoding/json/encode.go typeField func.
+func enumFields(t reflect.Type, opts MarshalOptions) []field {
+	// Fields to explore
+	current := []field{}
+	next := []field{{Type: t}}
+
+	// count of queued names
+	count := map[reflect.Type]int{}
+	nextCount := map[reflect.Type]int{}
+
+	visited := map[reflect.Type]struct{}{}
+	fields := []field{}
+
+	for len(next) > 0 {
+		current, next = next, current[:0]
+		count, nextCount = nextCount, map[reflect.Type]int{}
+
+		for _, f := range current {
+			if _, ok := visited[f.Type]; ok {
+				continue
+			}
+			visited[f.Type] = struct{}{}
+
+			for i := 0; i < f.Type.NumField(); i++ {
+				sf := f.Type.Field(i)
+				if sf.PkgPath != "" && !sf.Anonymous {
+					// Ignore unexported and non-anonymous fields
+					// unexported but anonymous field may still be used if
+					// the type has exported nested fields
+					continue
+				}
+
+				fieldTag := tag{}
+				fieldTag.parseAVTag(sf.Tag)
+				if opts.SupportJSONTags && fieldTag == (tag{}) {
+					fieldTag.parseJSONTag(sf.Tag)
+				}
+
+				if fieldTag.Ignore {
+					continue
+				}
+
+				ft := sf.Type
+				if ft.Name() == "" && ft.Kind() == reflect.Ptr {
```

Comment:
That is what i was thinking it seemed redundant, but didn't fully understand why the encoding/json code used this additional check.

After a bit of research I found the condition which it looks like this is catches. http://play.golang.org/p/2NThBfWeTU . Specifically if a type is defined as an alias for the pointer value of another type this code prevents the dereferencing too far. Interestingly enough I don't think this is replicated in the decoding side of encoding/json's indirect func. I'll add more tests to capture this case.